src/x86_64/service.c

```diff
@@ -103,6 +107,33 @@ closure_function(1, 2, void, fsstarted,
     closure_finish();
 }
 
+/* will become list I guess */
+static pagecache global_pagecache;
+
+/* This is very simplistic and uses a fixed drain threshold. This
+   should also take all cached data in system into account. For now we
+   just pick on the single pagecache... */
+
+#ifdef MM_DEBUG
+#define mm_debug(x, ...) do {rprintf("MM:   " x, ##__VA_ARGS__);} while(0)
+#else
+#define mm_debug(x, ...) do { } while(0)
+#endif
+void mm_service(void)
+{
+    if (!global_pagecache)
+        return;
+    heap p = (heap)heap_physical(&heaps);
+    u64 free = heap_total(p) - heap_allocated(p);
+    mm_debug("%s: total %ld, alloc %ld, free %ld\n", __func__, heap_total(p), heap_allocated(p), free);
+    if (free < CACHE_DRAIN_CUTOFF) {
+        u64 drain_bytes = CACHE_DRAIN_CUTOFF - free + PAGESIZE - 1;
```

Comment:
Shouldn't it be PAGESIZE_2M instead of PAGESIZE? Incidentally, I think this math should be done internally by pagecache_drain(), which should ensure that the number of page it drains (or tries to drain) covers at least `drain_bytes`.

Reply:
Agreed on both comments. Thanks.